File: project/autobahn/autobahn_python/autobahn.py
from enum import Enum
import io
from struct import Struct
from typing import Awaitable, Callable
import websockets
from websockets.server import ServerConnection
import asyncio

# Structure of the message:
# [flag] [length] [topic] [length] [payload]


class Flags(Enum):
    SUBSCRIBE = b"\x01"  # flag: 0x01
    UNSUBSCRIBE = b"\x02"  # flag: 0x02
    PUBLISH = b"\x03"  # flag: 0x03


class Autobahn:
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        self.websocket = None
        self.subscriptions = {}

    async def begin(self):
        self.websocket = await websockets.connect(f"ws://{self.host}:{self.port}")
        asyncio.create_task(self._listen())

    async def subscribe(self, topic: str, callback: Callable[[bytes], Awaitable[None]]):
        self.subscriptions[topic] = callback
        await self.__publish(Flags.SUBSCRIBE, topic.encode(), b"")

    async def unsubscribe(self, topic: str):
        del self.subscriptions[topic]
        await self.__publish(Flags.UNSUBSCRIBE, topic.encode(), b"")

    async def _listen(self):
        try:
            while True:
                try:
                    if self.websocket:
                        # Receive the message
                        msg = await self.websocket.recv()

                        # Ensure msg is bytes
                        if isinstance(msg, str):
                            msg = msg.encode("utf-8")

                        reader = io.BytesIO(msg)

                        _ = int.from_bytes(reader.read(1), byteorder="big")
                        topic_length = int.from_bytes(reader.read(2), byteorder="big")
                        topic = reader.read(topic_length).decode("utf-8")
                        payload = reader.read()

                        for subscription_topic, callback in self.subscriptions.items():
                            if topic.startswith(subscription_topic):
                                await callback(payload)
                except websockets.exceptions.ConnectionClosed:
                    break
        except Exception as e:
            print(f"Error in listener: {e}")

    async def __publish(self, flag: Flags, topic: bytes, message: bytes):
        if self.websocket:
            try:
                await self.websocket.send(
                    flag.value
                    + len(topic).to_bytes(2, "big")
                    + topic
                    + len(message).to_bytes(2, "big")
                    + message
                )
            except websockets.exceptions.ConnectionClosed:
                print("Connection closed")

    async def publish(self, topic: str, message: bytes):
        await self.__publish(Flags.PUBLISH, topic.encode(), message)

File: project/hybrid-frustum-pointnet/lidar/src/util/autobahn.cpp
#include "autobahn.h"

#include <iostream>
#include <stdexcept>

Autobahn::Autobahn(const std::string& host, const std::string& port)
    : host_(host), port_(port), resolver_(ioc_), ws_(ioc_) {}

Autobahn::~Autobahn() {
  stop();
  if (listening_thread_.joinable()) {
    listening_thread_.join();
  }
}

void Autobahn::begin() {
  // Resolve the hostname
  auto const results = resolver_.resolve(host_, port_);

  // Connect to the resolved endpoint
  asio::connect(ws_.next_layer(), results.begin(), results.end());

  // Perform the WebSocket handshake
  ws_.handshake(host_, "/");

  // Start the listening thread
  listening_thread_ = std::thread([this]() { listen(); });
}

void Autobahn::subscribe(const std::string& topic,
                         std::function<void(const std::string&)> callback) {
  subscriptions_[topic] = callback;
  send_message(Flags::SUBSCRIBE, topic, "");
}

void Autobahn::unsubscribe(const std::string& topic) {
  subscriptions_.erase(topic);
  send_message(Flags::UNSUBSCRIBE, topic, "");
}

void Autobahn::publish(const std::string& topic, const std::string& message) {
  send_message(Flags::PUBLISH, topic, message);
}

void Autobahn::listen() {
  try {
    while (running_) {
      beast::flat_buffer buffer;
      ws_.read(buffer);

      auto data = beast::buffers_to_string(buffer.data());
      parse_message(data);
    }
  } catch (const std::exception& e) {
    std::cerr << "Error in listener: " << e.what() << std::endl;
  }
}

void Autobahn::parse_message(const std::string& message) {
  try {
    auto it = message.begin();

    // Read flag
    Flags flag = static_cast<Flags>(*it++);

    // Read topic length
    uint16_t topic_length =
        static_cast<uint8_t>(*it++) << 8 | static_cast<uint8_t>(*it++);

    // Read topic
    std::string topic(it, it + topic_length);
    it += topic_length;

    // Read payload
    std::string payload(it, message.end());

    if (flag == Flags::PUBLISH) {
      for (const auto& [subscription_topic, callback] : subscriptions_) {
        if (topic.compare(0, subscription_topic.size(), subscription_topic) ==
            0) {
          callback(payload);
        }
      }
    }
  } catch (const std::exception& e) {
    std::cerr << "Error parsing message: " << e.what() << std::endl;
  }
}

void Autobahn::send_message(Flags flag, const std::string& topic,
                            const std::string& message) {
  try {
    // Create a binary buffer
    std::vector<uint8_t> buffer;

    // Add the flag (1 byte)
    buffer.push_back(static_cast<uint8_t>(flag));

    // Add the topic length (4 bytes, big-endian)
    uint32_t topic_length = topic.size();
    buffer.push_back(static_cast<uint8_t>((topic_length >> 24) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((topic_length >> 16) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((topic_length >> 8) & 0xFF));
    buffer.push_back(static_cast<uint8_t>(topic_length & 0xFF));

    // Add the topic (variable length)
    buffer.insert(buffer.end(), topic.begin(), topic.end());

    // Add the payload length (4 bytes, big-endian)
    uint32_t message_length = message.size();
    buffer.push_back(static_cast<uint8_t>((message_length >> 24) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((message_length >> 16) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((message_length >> 8) & 0xFF));
    buffer.push_back(static_cast<uint8_t>(message_length & 0xFF));

    // Add the payload (variable length)
    buffer.insert(buffer.end(), message.begin(), message.end());

    // Set WebSocket to binary mode
    ws_.binary(true);

    // Send the buffer directly as binary
    ws_.write(asio::buffer(buffer));
  } catch (const std::exception& e) {
    std::cerr << "Error sending message: " << e.what() << std::endl;
  }
}

void Autobahn::stop() {
  running_ = false;
  ws_.close(websocket::close_code::normal);
}